remove autosave preload and add per_page support

adamsilverstein opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

<!-- Insert a description of your changes here -->

Trac ticket: https://core.trac.wordpress.org/ticket/62057

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

adamsilverstein committed:
Remove autosaves preload, reducing memory footprint

adamsilverstein committed:
Merge branch 'trunk' into ticket/62057

adamsilverstein committed:
Try: add per_page support for autosaves

adamsilverstein committed:
Merge branch 'trunk' into ticket/62057

adamsilverstein committed:
add test for per_page parameter

adamsilverstein committed:
Add a test for the per_page parameter

adamsilverstein committed:
Add tests for autosaves per_page parameter

adamsilverstein committed:
phpcbf

adamsilverstein committed:
autosaves route test: add per_page to expected parameters

adamsilverstein committed:
preload a single autosave

adamsilverstein committed:
phpcbf

adamsilverstein committed:
merge trunk

adamsilverstein committed:
remove autosave preload entirely

TimothyBJacobs commented on src/wp-includes/rest-api/endpoints/class-wp-rest-autosaves-controller.php:
```diff
@@ -321,6 +334,12 @@ public function get_items( $request ) {
 
 		return rest_ensure_response( $response );
+
+		// Add pagination headers.
```
I think we'd want the Link headers as well.

adamsilverstein replied:
Ah, thanks for the feedback! I will investigate and add.

adamsilverstein committed:
Merge trunk into ticket/62057

adamsilverstein committed:
REST API: Allow the autosaves collection to be scoped by author.

WordPress stores at most one autosave per author, and the block editor only
ever reads the current user's, so returning every autosave for the post makes
the editor download a collection to discard all but one row of it.

Register an `author` parameter on the autosaves collection, shaped like the
one on the posts controller, and answer it with `wp_get_post_autosave()` for
each requested id. This replaces the `per_page` parameter tried earlier, which
narrowed the revision query before autosaves were filtered out of it and so
could return nothing at all on a post with revision history.

The block editor preload is scoped to match the path the editor requests,
rather than removed: the preload cache is keyed on the path, so an entry the
client never asks for is served to nobody.

See #62057.
